s3store/s3store_test.go

```diff
@@ -43,119 +43,129 @@ func TestCalcOptimalPartSize(t *testing.T) {
-			}
-		}
-		if len(err) > 0 {
-			t.Errorf(err)
+		equalparts = size / optimalPartSize
+		lastpartsize = size % optimalPartSize
+		assert.False(optimalPartSize < store.MinPartSize, fmt.Sprintf("Size %v, %v parts of size %v, lastpart %v: optimalPartSize < MinPartSize %v.\n", size, equalparts, optimalPartSize, lastpartsize, store.MinPartSize))
+		assert.False(optimalPartSize > store.MaxPartSize && calcError == nil, fmt.Sprintf("Size %v, %v parts of size %v, lastpart %v: optimalPartSize > MaxPartSize %v, but no error was returned.\n", size, equalparts, optimalPartSize, lastpartsize, store.MaxPartSize))
+		assert.False(size%optimalPartSize == 0 && equalparts > store.MaxMultipartParts, fmt.Sprintf("Size %v, %v parts of size %v, lastpart %v: more parts than MaxMultipartParts %v.\n", size, equalparts, optimalPartSize, lastpartsize, store.MaxMultipartParts))
+		assert.False(size%optimalPartSize > 0 && equalparts > store.MaxMultipartParts-1, fmt.Sprintf("Size %v, %v parts of size %v, lastpart %v: more parts than MaxMultipartParts %v.\n", size, equalparts, optimalPartSize, lastpartsize, store.MaxMultipartParts))
+		assert.False(lastpartsize > store.MaxPartSize, fmt.Sprintf("Size %v, %v parts of size %v, lastpart %v: lastpart > MaxPartSize %v.\n", size, equalparts, optimalPartSize, lastpartsize, store.MaxPartSize))
+		assert.False(lastpartsize > optimalPartSize, fmt.Sprintf("Size %v, %v parts of size %v, lastpart %v: lastpart > optimalPartSize %v.\n", size, equalparts, optimalPartSize, lastpartsize, optimalPartSize))
+		if debug {
+			fmt.Printf("Size %v, %v parts of size %v, lastpart %v, does exceed MaxObjectSize %v.\n", size, equalparts, optimalPartSize, lastpartsize, size > store.MaxObjectSize)
 		}
 	}
 	// fmt.Println("HighestApplicablePartSize", HighestApplicablePartSize)
 	// fmt.Println("RemainderWithHighestApplicablePartSize", RemainderWithHighestApplicablePartSize)
 }
 
+func TestAllPartSizes(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping test in short mode.")
+	}
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+	assert := assert.New(t)
+
+	s3obj := NewMockS3API(mockCtrl)
+	store := s3store.New("bucket", s3obj)
+
+	store.MinPartSize = 5
+	store.MaxPartSize = 5 * 1024
+	store.MaxMultipartParts = 1000
+	store.MaxObjectSize = store.MaxPartSize * store.MaxMultipartParts
+
+	var debug = false
+	var equalparts, lastpartsize int64
+
+	// sanity check
+	if store.MaxObjectSize > store.MaxPartSize*store.MaxMultipartParts {
+		t.Errorf("MaxObjectSize %v can never be achieved, as MaxMultipartParts %v and MaxPartSize %v only allow for an upload of %v bytes total.\n", store.MaxObjectSize, store.MaxMultipartParts, store.MaxPartSize, store.MaxMultipartParts*store.MaxPartSize)
+	}
+
+	for size := int64(1); size <= store.MaxObjectSize; size++ {
+		optimalPartSize, calcError := store.CalcOptimalPartSize(size)
```

Comment:
Is there a reason why you didn't check for `calcError == nil` here? For the range of `size` is should never return an error.

Reply:
Yes there is.
I view the range we are testing here as arbitrary and therefore the assertions should work with any range you choose. I actually would much prefer to test up to MaxObjectSize+1. An assertion `calcError == nil` would not work in this case.

Reply:
Also I believe that having identical assertions in the two tests concerning CalcOptimalPartSize helps prevent confusion. As we are testing the same thing, just different ranges of test cases, so why should the test differ.

Ideally, we would externalise the assertions into a function, which we use in both tests, which would assure, they stay the same over time. But as this would complicate things instead of making them more readable, I did refrain from trying that.

Reply:
I disagree. Call it preference or not, I am not a fan of having different scenarios into one test.

>  I actually would much prefer to test up to MaxObjectSize+1

I don't see it this way. This should be its own separate test since the expected outcome and the used assertions are different. Furthermore, this will make debugging issues easier since you can run a single test isolated using `go test -run`.

Reply:
Well, that is certainly a valid argument.

But if one would follow that perspective, the test would need to be ripped apart even further. Each one of the case statements in CalcOptimalPartSize then would represent its own test case in your scenario. 

Readability will be totally gone then. 

No single point to understand, what we are expecting from CalcOptimalPartSize anymore. And you would have to define universal scenarios for sizes, which will trigger the expected behaviour. Which will become rather complicated, when you consider, that MinPartSize, MaxpartSize, MaxMultipartParts and MaxObjectSize are variable too and might be very different in other S3 API implementations.

You might end up with a lot more tests, than you are prepared to handle.

Reply:
I believe there is a difference between testing something like WriteChunk, which by itself depends on other APIs, functions and methods and a helper function with only one input and one output and no additional level of abstraction.

For WriteChunk the quality of the data on the input can differ quite significantly, both in value, but also in its structural composition. And moreover, we depend on other functions, methods and interfaces below it.

While CalcOptimalPartSize just takes a number and will return another one, or an error. But it will not venture off into other layers of abstraction.